### PR TITLE
OP-717 [OH-CORE] DICOM BLOB in a separate table

### PR DESCRIPTION
**Pull request by VaheAvetikyan**
See OP-717 

- move the BLOB field from DICOM to a separate table

### CODE REVIEW

**Comment by VaheAvetikyan** [comment]
Hello,
My first PR.
I mainly followed the implementation of PATIENT_PROFILE_PHOTO as suggested in the issue description.

File: src/main/java/org/isf/dicom/model/DicomData.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+@Entity
+@Table(name = "OH_DICOM_DATA")
+public class DicomData {
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(DicomData.class);
+
+	@Id
+	@GeneratedValue(strategy = GenerationType.AUTO)
+	@Column(name = "DICOM_DATA_ID")
+	private Integer code;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/dicom/model/DicomData.java:53]: I'm always curious as to why the primitive 'int` is not used for `Integer`

File: sql/step_91_add_dicom_data_table.sql [hunk 2]
```diff
@@ -0,0 +1,30 @@
+-- add foreign key from dicom to dicom_data
+-- this means that the dicom table will be responsible in maintaining the relationship between dicom and the dicom data table
+ALTER TABLE DICOM ADD COLUMN DICOM_DATA_ID int(11);
+
+-- add the new table
+-- note that:
+-- 1. adding primary key with auto increment. We will not be setting this auto increment manually
+-- 2. adding TEMP_DATA_ID for migration purpose, so that we can relate existing data from DICOM table to this new table while doing migration
+CREATE TABLE DICOM_DATA (
```
Comment (dbmalkovsky) [inline sql/step_91_add_dicom_data_table.sql:9]: All tables are prefixed by `OH_`

File: sql/step_91_add_dicom_data_table.sql [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+                                       DICOM_DATA_ID int NOT NULL AUTO_INCREMENT,
```
Comment (dbmalkovsky) [inline sql/step_91_add_dicom_data_table.sql:10]: all columns in a table are prefixed by some abbreviation of the table name , some something like `DD.DICOM_DATA_IT`

Comment (VaheAvetikyan) [reply sql/step_91_add_dicom_data_table.sql]: Added prefix DMD_

**Comment by mwithi** [comment]
> Hello, My first PR. I mainly followed the implementation of PATIENT_PROFILE_PHOTO as suggested in the issue description.

Hello @VaheAvetikyan, thanks a lot for the PR. Please could you check why build is failing? ([link](https://github.com/informatici/openhospital-core/actions/runs/4518548831/jobs/7970075230?pr=919))

File: sql/step_91_add_dicom_data_table.sql [hunk 4]
```diff
@@ -0,0 +1,30 @@
+-- add foreign key from dicom to dicom_data
+-- this means that the dicom table will be responsible in maintaining the relationship between dicom and the dicom data table
+ALTER TABLE OH_DICOM ADD COLUMN DMD_DATA_ID int(11);
```
Comment (mwithi) [inline sql/step_91_add_dicom_data_table.sql:3]: in the OH_DICOM table prefixes are all DM_

Comment (VaheAvetikyan) [reply sql/step_91_add_dicom_data_table.sql]: changed

File: sql/step_91_add_dicom_data_table.sql [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+-- update references from DICOM table back to OH_DICOM_DATA_ID
+UPDATE OH_DICOM
+    INNER JOIN OH_DICOM_DATA
+    ON OH_DICOM.DM_FILE_ID = OH_DICOM_DATA.TEMP_DATA_ID
+SET OH_DICOM.DMD_DATA_ID=OH_DICOM_DATA.DMD_DATA_ID;
+
+-- drop the temporary column
+ALTER TABLE OH_DICOM_DATA DROP TEMP_DATA_ID;
+ALTER TABLE OH_DICOM DROP DM_DATA;
```
[the remark this answers is not in the document]
Comment (VaheAvetikyan) [reply sql/step_91_add_dicom_data_table.sql]: applied

File: sql/step_04_all_following_steps.sql [hunk 6]
```diff
@@ -84,4 +84,5 @@ source step_87_alter_table_opd.sql;
 source step_88_alter_table_patient.sql;
 source step_89_alter_tables_innodb.sql;
 source step_90_login_retries.sql;
+source step_91_add_dicom_data_table.sql;
```
Comment (dbmalkovsky) [inline sql/step_04_all_following_steps.sql:87]: Please renumber; there should not be 2 `step_91_`

File: sql/step_92_add_dicom_data_table.sql [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    (SELECT d.DM_FILE_ID, d.DM_DATA
+     FROM OH_DICOM d
+     WHERE d.DM_DATA IS NOT NULL);
+
+-- drop original column
+ALTER TABLE OH_DICOM DROP DM_DATA;
+
+-- add DMD_FILE_ID foreign key
+ALTER TABLE OH_DICOM_DATA
+    ADD CONSTRAINT FK_DMD_FILE_ID
```
Comment (mwithi) [inline sql/step_92_add_dicom_data_table.sql:21]: we use format `FK_<this table>_<other table>` without prefix `OH_`:

```suggestion
    ADD CONSTRAINT FK_DICOM_DATA_DICOM
```

**Review by mwithi — APPROVED** [review]
changed FK name by myself!
thanks a lot!

**Pull request merged**